cpp/src/parquet/row_ranges.h

```diff
@@ -0,0 +1,201 @@
+   * the list. The specified range shall be larger(*) than the last one or might be
+   * overlapped with some of the last ones.
+   * (*) [a, b] < [c, d] if b < c
+   */
+  void Add(const Range& range) {
+    CheckRangeIsValid(range);
+    auto& range_to_add = const_cast<Range&>(range);
+    for (int i = static_cast<int>(ranges_.size() - 1); i >= 0; --i) {
+      Range& last = ranges_[i];
+      if (last.IsAfter(range)) {
+        throw ParquetException("Ranges should be in ascending order");
+      }
+      const Range& u = Range::Union(last, range_to_add);
+      if (u == EMPTY_RANGE) {
+        break;
+      }
+      range_to_add = u;
+      ranges_.pop_back();
+    }
+    ranges_.push_back(range_to_add);
+  }
+
+  const std::vector<Range>& GetRanges() const { return ranges_; }
+
+  const Range& GetRange(int i) const {
+    if (i >= static_cast<int>(ranges_.size())) {
+      throw ParquetException("Index overflow");
+    }
+    return ranges_[i];
+  }
+
+  int64_t GetRowCount() const {
+    int64_t row_count = 0;
+    for (const Range& range : ranges_) {
+      row_count += range.RowCount();
+    }
+    return row_count;
+  }
+
+  bool IsOverlapping(int64_t from, int64_t to) const;
+
+  /**
+   * Calculates the intersection of the two specified RowRanges object. Two ranges
+   * intersect if they have common elements otherwise the result is empty. <pre> For
+   * example: [113, 241] ∩ [221, 340] = [221, 241] while [113, 230] ∩ [231, 340] =
+   * &lt;EMPTY&gt;
+   * </pre>
+   * @param left left RowRanges
+   * @param right right RowRanges
+   * @return a mutable RowRanges contains all the row indexes that were contained in both
+   * of the specified objects
+   */
+  static RowRanges Intersection(const RowRanges& left, const RowRanges& right) {
+    const auto& left_ranges = left.GetRanges();
+    const auto& right_ranges = right.GetRanges();
+    const int num_right_ranges = static_cast<int>(right_ranges.size());
+
+    RowRanges result;
+    int right_index = 0;
+    for (const Range& l : left_ranges) {
+      for (int i = right_index, n = num_right_ranges; i < n; ++i) {
+        Range r = right_ranges[i];
+        if (l.IsBefore(r)) {
+          break;
+        } else if (l.IsAfter(r)) {
+          right_index = i + 1;
+          continue;
+        }
+        result.Add(Range::Intersection(l, r));
+      }
+    }
+
+    return result;
+  }
+
+  bool operator==(const RowRanges& other) const { return ranges_ == other.ranges_; }
+
+  bool operator!=(const RowRanges& other) const { return !(*this == other); }
+
+  std::string ToString() const {
+    std::stringstream ss;
+    ss << "[";
+    for (size_t i = 0; i < ranges_.size(); i++) {
+      if (i == 0) {
+        ss << ranges_[i].ToString();
+      } else {
+        ss << "," << ranges_[i].ToString();
+      }
+    }
+    ss << "]";
+    return ss.str();
+  }
+
+  static constexpr Range EMPTY_RANGE{std::numeric_limits<int64_t>::max(),
```

Comment:
shouldn't this be something like {0, 0}, as it stands now it is an invalid range?  Also kEmptyRange I think is preferred in new code

Reply:
> shouldn't this be something like {0, 0}, as it stands now it is an invalid range

``EMPTY_RANGE ``  represents InvalidRange or EmptyRange in different scenes. ``{0, 0}`` means hit the first row, so we can't use it to mean empty range. ``kEmptyRange`` is more appropriate, I will modify it later.